sys/pm_layered/pm.c

```diff
@@ -81,7 +83,10 @@ void pm_unblock(unsigned mode)
 
 pm_blocker_t pm_get_blocker(void)
 {
-    return pm_blocker;
+    unsigned state = irq_disable();
+    pm_blocker_t result = pm_blocker;
+    irq_restore(state);
+    return result;
```

Comment:
This should only be necessary on non-32 bit machines, no?

Reply:
There sadly is no guarantee, that the compiler will generate a single read whenever possible. There were examples on GCC generating two 32 bit stores on x86_64 when writing to a `volatile`, as two stores with an immediate where faster than preparing the correct value in a register and writing that back to memory. As a result, some linux drivers stopped working when compiled with that GCC version. I think it is safe to say that the Linux kernel guys convinced the compiler guys that their compilers have no host to run on if they break kernels by splitting a single `volatile` accesses into two accesses. But for non-`volatile` accesses we should still be careful to do any assumptions on how the generated assembly looks like.

So what we could IMO safely do is something like this:

```C
pm_blocker_t pm_get_blocker(void)
{
    pm_blocker_t result;
    if (ARCH_32BIT) {
        volatile uint32_t *tmp = &pm_blocker.val_u32;
        result.val_u32 = *tmp;
    }
    else {
        unsigned state = irq_disable();
        result = pm_blocker;
        irq_restore(state);
    }
    return result;
}
```

It would be nice to make the `FEATURES` like `arch32_bit` available to C code for this. The compiler should than drop the dead branch.

Reply:
`ARCH_32_BIT` is now available, can you adapt to your suggestion?

Reply:
ping @maribu  :)

Reply:
Sorry for the delay. I still had https://github.com/RIOT-OS/RIOT/pull/14331 in the pipeline, which is a better solution than what I suggested above.

I'm not sure if we should just stall this PR until https://github.com/RIOT-OS/RIOT/pull/14331 is merged, or get it in without the optimization for now.

Reply:
I would merge without optimizations, that PR can take time and this is still potential source for issues.